runtime/autoload/man.vim

```diff
@@ -11,127 +22,207 @@ catch /E145:/
-    exec "let s:man_tag_lin=s:man_tag_lin_".s:man_tag_depth
-    exec "let s:man_tag_col=s:man_tag_col_".s:man_tag_depth
-    exec s:man_tag_buf."b"
-    exec s:man_tag_lin
-    exec "norm! ".s:man_tag_col."|"
-    exec "unlet s:man_tag_buf_".s:man_tag_depth
-    exec "unlet s:man_tag_lin_".s:man_tag_depth
-    exec "unlet s:man_tag_col_".s:man_tag_depth
-    unlet s:man_tag_buf s:man_tag_lin s:man_tag_col
+function! man#complete(ArgLead, CmdLine, CursorPos) abort
+  let args = split(a:CmdLine)
+  let l = len(args)
+  " if the cursor (|) is at ':Man printf(|' then
+  " make sure to display the section. See s:get_candidates
+  let fpage = 0
+  " if already completed a manpage, we return
+  if (l > 1 && args[1] =~# ')\f*$') || l > 3
+    return
+  elseif l == 3
+    " cursor (|) is at ':Man 3 printf |'
+    if empty(a:ArgLead)
+      return
+    endif
+    let sect = tolower(args[1])
+    let page = a:ArgLead
+  elseif l == 2
+    " cursor (|) is at ':Man 3 |'
+    if empty(a:ArgLead)
+      let page = ''
+      let sect = tolower(args[1])
+    elseif a:ArgLead =~# '^\f\+(\f*$'
+      " cursor (|) is at ':Man printf(|'
+      let tmp = split(a:ArgLead, '(')
+      let page = tmp[0]
+      let sect = tolower(substitute(get(tmp, 1, ''), ')$', '', ''))
+      let fpage = 1
+    else
+      " cursor (|) is at ':Man printf|'
+      let page = a:ArgLead
+      let sect = ''
+    endif
+  else
+    let page = ''
+    let sect = ''
   endif
+  return s:get_candidates(page, sect, fpage)
 endfunction
 
-" Expects a string like 'access' or 'access(2)'.
-function s:parse_page_and_section(sect, str) abort
-  try
-    let [page, sect] = matchlist(a:str, '\v\C([-.[:alnum:]_]+)%(\(([-.[:alnum:]_]+)\))?')[1:2]
-    if empty(sect)
-      let sect = a:sect
+function! s:get_candidates(page, sect, fpage) abort
+  let candidates = globpath(s:MANDIRS(),'*/'.a:page.'*.'.a:sect.'*', 0, 1)
+  let find = '\(.\+\)\.\%('.s:man_extensions.'\)\@!\'
+  " if the page is a path, complete files
+  if empty(a:sect) && a:page =~# '\/'
+    "TODO(nhooyr) why does this complete the last one automatically
+    let candidates = glob(a:page.'*', 0, 1)
+  else
+    " if the section is not empty and the cursor (|) is not at
+    " ':Man printf(|' then do not show sections.
+    if !empty(a:sect) && !a:fpage
+      let find .= '%([^.]\+\).*'
+      let repl = '\1'
+    else
+      let find .= '([^.]\+\).*'
+      let repl = '\1(\2)'
     endif
-  catch
-    echoerr 'man.vim: failed to parse: "'.a:str.'"'
-  endtry
-
-  return [page, sect]
-endfunction
-
-function s:cmd(sect, page) abort
-  if !empty(a:sect)
-    return s:man_sect_arg.' '.a:sect.' '.a:page
+    call map(candidates, 'substitute((fnamemodify(v:val, ":t")), find, repl, "")')
   endif
-  return a:page
+  return candidates
 endfunction
 
-function s:find_page(sect, page) abort
-  let where = system('/usr/bin/man '.s:man_find_arg.' '.s:cmd(a:sect, a:page))
-  return (where =~# '^ */')
+function! s:MANDIRS() abort
```

Comment:
Same comment about consistency.  Also, if you don't plan to use this function elsewhere, it might be better to put the body in the `s:get_candidates()` function since it's pretty short.